core/unix/loader.c

```diff
@@ -1767,6 +1759,14 @@ privload_early_inject(void **sp, byte *old_libdr_base, size_t old_libdr_size)
         takeover_ptrace((ptrace_stack_args_t *)sp);
     }
 
+    kernel_init_sp = (void *)sp;
+
+    /* XXX i#47: for Linux, we can't easily have this option on by default as
+     * code like get_application_short_name() called from drpreload before
+     * even _init is run needs to have a non-early default.
+     */
+    dynamo_options.early_inject = true;
```

Comment:
Are we sure this option should be set when using ptrace?  Even when we created the app's process, we did an execve on the app and so it was loaded with the system loader, which ran a bunch of code already.  I suppose we're loading libdynamorio, rather than the system loader, which is one major use of -early_inject in the code.  Does any code assume it also means we've seen all app code?

Reply:
I suppose my question isn't soley about your PR here but about the ptrace option in general.  But, your description says "So we consider ptraced child is not early_inject", yet you're still setting early_inject?  I'm confused.

Reply:
OK, I think I see: takeover_ptrace does not return, so control never reaches down here.  Could you add an `ASSERT_NOT_REACHED();` line after the takeover_ptrace call?  That will make it clear and avoid my initial confusion.